IBM/ubiquity
# Improve performance for mpath Discovery method

Improve performance by skip discover devices by sg_inq during the attach flow (create POD).

**Description:**
In case you have many devices on the worker nodes and you create a new POD with PVC, before this fix the discover would go on all the PVCs devices on the node and trigger sg_inq on each one of them in order to search for the device for the new PVC (wwn). Well in most use cases there is no need to do this overhead during attach since the device probably doesn't exist on the host or if it is then it should appear in multipath -ll already. So to improve performance we decided to skip this deepDiscovery(by sg_inq on all the devices). And it will already be covered anyway after the rescanAll.

**Improve discovery performance by applying the following things:**
1. Do NOT do sg_inq deepDiscovery during the RescanAll in the attach flow.
   Its waist of time to do deepDiscovery in this case, since its already done after the RescanAll when discovering the new device.

2. Do sg_inq deepDiscovery only in the following cases:
  2.1. During the attach after the RescanAll
  2.2. When Discover for cleanup(unmount flow).

3. Remove double sg_inq validation if discover during deepDiscovery

4. Remove Stat(mpath) validation during Discover, since its already validated during GetWwnByScsiInq() or DiscoverBySgInq()

5. Fix interfaces, UT and fakes according (Still need to add more UT)


**In addition** this PR improve the DiscoverBySgInq by **searching only IBM devices**(devices with "IBM" vendor tag presented in multipath -ll). A side affect of this filtering is that it will not work on faulty devices that usually doesn't present any Vendor tag on.


**How to test**
1. create POD1 with PVC1 and check in the flex logs that the volume didn't had to discover using DiscoverBySgInq, means you should see the message "mpath device was NOT found for WWN [%s] in multipath -ll"(where %s is the wwn of the PVC). but you should not see any try to run sg_inq on this WWN (debug for GetWwnByScsiInq with this mpath device)

2. Try to get in multipath -ll with bad devices, then run create POD2 with PVC2, this this case there is high chance that multipath -ll will not present the right WWN for the new PVC2 device.  if so then you should see in the log that flex tried to identify this PVC2 device by using sg_inq. You can see in the log message like "Warning: device [%s] found for WWN [%s] after running sg_inq"


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ibm/ubiquity/171)
